### PR TITLE
fix: address memory leaks caused by celery

### PR DESCRIPTION
**Pull request by lukmccall**
### Why

Fixes memory leaks.
Close FREGE-160.

### Background

https://medium.com/@aaron.reyna/python-celery-ram-intensive-tasks-and-a-memory-leak-c2681ee98c9
https://chase-seibert.github.io/blog/2013/08/03/diagnosing-memory-leaks-python.html

TL;DR
Long-running Python jobs that consume a lot of memory while running may not return that memory to the operating system until the process actually terminates, even if everything is garbage collected properly. That was news to me, but it’s true. What this means is that processes that do need to use a lot of memory will exhibit a “high water” behaviour, where they remain forever at the level of memory usage that they required at their peak.

### Solutions

- Make `crawl_repos_task` spawn another task instead of an inner infinite loop.
- Set `worker_max_tasks_per_child` to ensure celery will release resources. 

### Test plan

Run the whole application for about 45 min and check if the memory consumption stays at a similar level.
image

### Future improvement

We may want to replace GitPython with something else (like https://www.pygit2.org/repository.html), cause it may leak. 
According to the documentation (https://github.com/gitpython-developers/GitPython#leakage-of-system-resources):
> GitPython is not suited for long-running processes (like daemons) as it tends to leak system resources. It was written in a time where destructors (as implemented in the __del__ method) still ran deterministically.

Right now, it isn't really a problem. The operating system will clean those leaks shortly after the celery kills the worker process.

### CODE REVIEW

**Comment by lukmccall** [comment]
@ConsoleRunner, @vladfedoriuk what do you think?

File: backend/fregepoc/celery_app.py [hunk 1]
```diff
@@ -5,7 +5,7 @@
 # set the default Django settings module for the 'celery' program.
 os.environ.setdefault("DJANGO_SETTINGS_MODULE", "fregepoc.settings")
 
-app = Celery("frege")
+app = Celery("frege", worker_max_tasks_per_child=4)
```
Comment (vladfedoriuk) [inline backend/fregepoc/celery_app.py:8]: Wouldn't it be more appropriate to read this parameter from env vars? What do you think?

File: backend/fregepoc/repositories/tasks.py [hunk 2]
```diff
@@ -37,15 +37,16 @@ def init_worker(**kwargs):
...
+        crawl_repos_task.apply_async(args=(indexer_cls.__name__,))
 
 
 @shared_task
 def crawl_repos_task(indexer_class_name):
     indexer_model = apps.get_model("indexers", indexer_class_name)
     indexer: BaseIndexer = indexer_model.load()
-    for repo in indexer:
+
+    batch = next(iter(indexer))
```
Comment (vladfedoriuk) [inline backend/fregepoc/repositories/tasks.py:48]: Could you lease elaborate a bit more on why this change is necessary? It seems to me that this particular celery task does not have anything to do with file descriptors.

Comment (vladfedoriuk) [inline backend/fregepoc/repositories/tasks.py:48]: Here I would rather give a default parameter: `batch = next(iter(indexer), [])`. This will prevent from `StopIteration` error being raised and not handled.

File: backend/fregepoc/repositories/tasks.py [hunk 3]
```diff
@@ -60,6 +61,8 @@ def crawl_repos_task(indexer_class_name):
             args=(indexer_class_name,),
             countdown=indexer.rate_limit_timeout.seconds,
         )
+    else:
```
Comment (vladfedoriuk) [inline backend/fregepoc/repositories/tasks.py:64]: IMO `for...else` structure is not a good practice. Perhaps, `if not batch` will be more readable?

File: backend/fregepoc/repositories/tasks.py [hunk 4]
```diff
@@ -104,6 +107,8 @@ def process_repo_task(repo_pk):
 
     for repo_file in repo_files:
         analyze_file_task.apply_async(args=(repo_file.pk,))
+
+    repo_obj.close()
```
Comment (vladfedoriuk) [inline backend/fregepoc/repositories/tasks.py:111]: I suggest using `contextlib.closing` context manager instead: https://docs.python.org/3/library/contextlib.html#contextlib.closing instead as it basically closes objects in a `final` statement ensuring us that the resource will have gotten closed even if some kind of error has occurred.

File: backend/fregepoc/repositories/tasks.py [hunk 5]
```diff
@@ -104,6 +107,8 @@ def process_repo_task(repo_pk):
...
     _finalize_repo_analysis(repo)
```
Comment (vladfedoriuk) [inline backend/fregepoc/repositories/tasks.py:114]: Please, remove redundant empty lines at the end of the files. Have you already installed `pre-commit` hooks? If you install them, they will be doing this kind of stuff for you.

File: backend/fregepoc/repositories/tasks.py (hunk 3 above)
Comment (lukmccall) [reply backend/fregepoc/repositories/tasks.py:64]: I didn't use `for...else`. It may look like that. However, this `else` is connected with `if` above.

File: backend/fregepoc/repositories/tasks.py (hunk 2 above)
Comment (lukmccall) [reply backend/fregepoc/repositories/tasks.py]: Ye, but this task is doing a lot of requests which also may leave some garbage. Without splitting it, celery will leak some system resources. Also, to my knowledge, this framework prefers smaller but shorter tasks rather than longer ones.

File: backend/fregepoc/repositories/tasks.py (hunk 3 above)
Comment (vladfedoriuk) [reply backend/fregepoc/repositories/tasks.py:85]: Oh, okay, sorry, my bad 😅

**Comment by lukmccall** [comment]
@vladfedoriuk, @ConsoleRunner can we merge it?

**Comment by ConsoleRunner** [comment]
All conversations were resolved, everything seems ok; I will merge it asap - unless @vladfedoriuk wants to add something.

**Comment by vladfedoriuk** [comment]
Sorry for not replying on time. Everything looks good to me. :)
